src/main.rs

```diff
@@ -38,13 +52,54 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let addr = "0.0.0.0:4243".parse().unwrap();
     tracing::info!("chariott listening on {addr}");
 
-    let router = Server::builder()
-        .add_service(ChariottServiceServer::new(ChariottServer::new(registry, broker)));
+    let server = Arc::new(ChariottServer::new(registry, broker));
+    let router =
+        Server::builder().add_service(ChariottServiceServer::from_arc(Arc::clone(&server)));
 
     #[cfg(build = "debug")]
     let router = router.add_service(reflection_service);
 
-    router.serve_with_ctrl_c_shutdown(addr).await?;
+    let error_cancellation_token = CancellationToken::new();
+    let ctrl_c_cancellation_token = ctrl_c_cancellation();
+
+    let prune_loop = {
+        let ctrl_c_cancellation_token = ctrl_c_cancellation_token.clone();
+        let error_cancellation_token = error_cancellation_token.child_token();
+
+        async move {
+            tracing::debug!("Prune loop running (TTL = {registry_entry_ttl:?}).");
+            loop {
+                let wakeup_deadline = server.registry_do(|reg| {
+                    reg.prune(Instant::now()).unwrap_or(Instant::now() + registry_entry_ttl)
```

Comment:
Why not use unwrap_or_else to lazy evaluate?

Reply:
My initial thinking was because it's not expensive to compute and I wanted to avoid a closure with `Instant::now()`, but digging into this now, I see that everything gets in-lined without any closure. Here's a simple test I tried:

```rust
use std::time::*;

pub fn test1() -> Instant {
    let now = Instant::now();
    prune(now).unwrap_or(now + Duration::from_secs(5))
}

pub fn test2() -> Instant {
    let now = Instant::now();
    prune(now).unwrap_or_else(|| now + Duration::from_secs(5))
}

#[inline(never)]
pub fn prune(time: Instant) -> Option<Instant> {
    None
}
```

which gets compiled to the following:

```asm
example::test1:
        push    r14
        push    rbx
        sub     rsp, 24
        call    qword ptr [rip + std::time::Instant::now@GOTPCREL]
        mov     r14, rax
        mov     rbx, rdx
        mov     rdi, rsp
        call    qword ptr [rip + example::prune@GOTPCREL]
        mov     edx, 5
        mov     rdi, r14
        mov     rsi, rbx
        xor     ecx, ecx
        call    qword ptr [rip + <std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add@GOTPCREL]
        cmp     qword ptr [rsp], 0
        je      .LBB0_2
        mov     rdx, qword ptr [rsp + 16]
        mov     rax, qword ptr [rsp + 8]
.LBB0_2:
        add     rsp, 24
        pop     rbx
        pop     r14
        ret

example::test2:
        push    r14
        push    rbx
        sub     rsp, 24
        call    qword ptr [rip + std::time::Instant::now@GOTPCREL]
        mov     r14, rax
        mov     rbx, rdx
        mov     rdi, rsp
        call    qword ptr [rip + example::prune@GOTPCREL]
        cmp     qword ptr [rsp], 0
        je      .LBB1_1
        mov     rax, qword ptr [rsp + 8]
        mov     rdx, qword ptr [rsp + 16]
        jmp     .LBB1_3
.LBB1_1:
        mov     edx, 5
        mov     rdi, r14
        mov     rsi, rbx
        xor     ecx, ecx
        call    qword ptr [rip + <std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add@GOTPCREL]
.LBB1_3:
        add     rsp, 24
        pop     rbx
        pop     r14
        ret
```

One can see in `test2` that after the call to `prune`:

```asm
        call    qword ptr [rip + example::prune@GOTPCREL]
```

it checks for `None`:

```asm
        cmp     qword ptr [rsp], 0
```

and [if it's the case, then simply _jumps_](https://en.wikibooks.org/wiki/X86_Assembly/Control_Flow#Jump_if_Equal) to a label:

```asm
        je      .LBB1_1
```

The code at that label then does the `now + Duration::from_secs(5)` bit:

```asm
.LBB1_1:
        mov     edx, 5
        mov     rdi, r14
        mov     rsi, rbx
        xor     ecx, ecx
        call    qword ptr [rip + <std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add@GOTPCREL]
```

I've therefore gone ahead and used `unwrap_or_else` instead with 09607fda01080ac3b8d3d1094ecd0316f94654eb.